hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/kac/TestApacheClientConnectionPool.java

```diff
@@ -0,0 +1,129 @@
+  }
+
+  @Test
+  public void testSysPropAppliedPool() throws IOException {
+    final String customPoolSize = "10";
+    System.setProperty(HTTP_MAX_CONN_SYS_PROP, customPoolSize);
+    validatePoolSize(Integer.parseInt(customPoolSize));
+  }
+
+  private void validatePoolSize(int size) throws IOException {
+    KeepAliveCache keepAliveCache = KeepAliveCache.getInstance();
+    final HttpRoute routes = new HttpRoute(new HttpHost("localhost"));
+    final HttpClientConnection[] connections = new HttpClientConnection[size * 2];
+
+    for (int i = 0; i < size * 2; i++) {
+      connections[i] = Mockito.mock(HttpClientConnection.class);
+    }
+
+    for (int i = 0; i < size * 2; i++) {
+      keepAliveCache.put(routes, connections[i]);
+    }
+
+    for (int i = size; i < size * 2; i++) {
+      Mockito.verify(connections[i], Mockito.times(1)).close();
+    }
+
+    for (int i = 0; i < size * 2; i++) {
+      if (i < size) {
+        Assert.assertNotNull(keepAliveCache.get(routes));
+      } else {
+        Assert.assertNull(keepAliveCache.get(routes));
+      }
+    }
+    System.clearProperty(HTTP_MAX_CONN_SYS_PROP);
+    keepAliveCache.close();
+  }
+
+  @Test
+  public void testKeepAliveCache() throws IOException {
+    KeepAliveCache keepAliveCache = KeepAliveCache.getInstance();
+    final HttpRoute routes = new HttpRoute(new HttpHost("localhost"));
+    HttpClientConnection connection = Mockito.mock(HttpClientConnection.class);
+
+    keepAliveCache.put(routes, connection);
+
+    Assert.assertNotNull(keepAliveCache.get(routes));
+    keepAliveCache.put(routes, connection);
+
+    final HttpRoute routes1 = new HttpRoute(new HttpHost("localhost1"));
+    Assert.assertNull(keepAliveCache.get(routes1));
+    keepAliveCache.close();
+  }
+
+  @Test
+  public void testKeepAliveCacheCleanup() throws Exception {
+    KeepAliveCache keepAliveCache = KeepAliveCache.getInstance();
+    final HttpRoute routes = new HttpRoute(new HttpHost("localhost"));
+    HttpClientConnection connection = Mockito.mock(HttpClientConnection.class);
+    keepAliveCache.put(routes, connection);
+
+    Thread.sleep(2 * KAC_CONN_TTL);
+    Mockito.verify(connection, Mockito.times(1)).close();
+    Assert.assertNull(keepAliveCache.get(routes));
+    Mockito.verify(connection, Mockito.times(1)).close();
+    keepAliveCache.close();
+  }
+
+  @Test
+  public void testKeepAliveCacheCleanupWithConnections() throws Exception {
+    KeepAliveCache keepAliveCache = KeepAliveCache.getInstance();
+    keepAliveCache.pauseThread();
+    final HttpRoute routes = new HttpRoute(new HttpHost("localhost"));
+    HttpClientConnection connection = Mockito.mock(HttpClientConnection.class);
+    keepAliveCache.put(routes, connection);
+
+    Thread.sleep(2 * KAC_CONN_TTL);
+    Mockito.verify(connection, Mockito.times(0)).close();
+    Assert.assertNull(keepAliveCache.get(routes));
+    Mockito.verify(connection, Mockito.times(1)).close();
+    keepAliveCache.close();
+  }
+}
```

Comment:
1. You can add a test case where multiple threads are trying to put and get from the cache simultaneously and verify the behaviour.
2. If we retrieve a connection from the cache and then put it back, will it use the same instance ?
3. If I put more connections than cache can hold what happens to the remaining connections can be verified ?

Reply:
1. Added new test which would do:
    - Verify KAC behavior for multiple route in parallel
    - Verify KAC behavior for one route, for which put and get methods are getting called in parallel.
2. Yes. Added a test to verify recache.
3. Yes, already there.